File: backend/services/pathfinding-service/src/main/java/com/navigation/pathfinding/domain/PathfindingAlgorithms.java
package com.navigation.pathfinding.domain;

import java.util.Set;

public enum PathfindingAlgorithms {
  A_STAR(PathfindingOptimizations.ALL_OPTIMIZING),
  BIDIRECTIONAL_A_STAR(PathfindingOptimizations.ALL_OPTIMIZING),
  BELLMAN_FORD(PathfindingOptimizations.ALL_OPTIMIZING),
  BFS(Set.of(PathfindingOptimizations.NUMBER_OF_NODES)),
  BIDIRECTIONAL_BFS(Set.of(PathfindingOptimizations.NUMBER_OF_NODES)),
  DIJKSTRA(PathfindingOptimizations.ALL_OPTIMIZING),
  BIDIRECTIONAL_DIJKSTRA(PathfindingOptimizations.ALL_OPTIMIZING),
  DFS(Set.of(PathfindingOptimizations.NONE)),
  GREEDY_BEST_FIRST_SEARCH(PathfindingOptimizations.COORDINATES_BASED),
  BIDIRECTIONAL_GREEDY_BEST_FIRST_SEARCH(PathfindingOptimizations.COORDINATES_BASED);

  public final Set<PathfindingOptimizations> SUPPORTED_OPTIMIZATIONS;

  PathfindingAlgorithms(Set<PathfindingOptimizations> SUPPORTED_OPTIMIZATIONS) {
    this.SUPPORTED_OPTIMIZATIONS = SUPPORTED_OPTIMIZATIONS;
  }

  public boolean isOptimizationAllowed(PathfindingOptimizations optimization) {
    return SUPPORTED_OPTIMIZATIONS.contains(optimization);
  }
}

File: client/src/constants/pathfinding-algorithms.ts
export enum PathfindingAlgorithmTypes {
  DIJKSTRA = 'DIJKSTRA',
  A_STAR = 'A_STAR',
  BIDIRECTIONAL_A_STAR = 'BIDIRECTIONAL_A_STAR',
  BELLMAN_FORD = 'BELLMAN_FORD',
  BFS = 'BFS',
  BIDIRECTIONAL_BFS = 'BIDIRECTIONAL_BFS',
  BIDIRECTIONAL_DIJKSTRA = 'BIDIRECTIONAL_DIJKSTRA',
  DFS = 'DFS',
  GREEDY_BEST_FIRST_SEARCH = 'GREEDY_BEST_FIRST_SEARCH',
  BIDIRECTIONAL_GREEDY_BEST_FIRST_SEARCH = 'BIDIRECTIONAL_GREEDY_BEST_FIRST_SEARCH',
}

export const ALGORITHM_TYPE_TRANSLATIONS = {
  [PathfindingAlgorithmTypes.DIJKSTRA]: 'Dijkstra',
  [PathfindingAlgorithmTypes.BFS]: 'BFS',
  [PathfindingAlgorithmTypes.BIDIRECTIONAL_BFS]: 'Bidirectional BFS',
  [PathfindingAlgorithmTypes.BELLMAN_FORD]: 'Bellman-Ford',
  [PathfindingAlgorithmTypes.A_STAR]: 'A*',
  [PathfindingAlgorithmTypes.BIDIRECTIONAL_A_STAR]: 'Bidirectional A*',
  [PathfindingAlgorithmTypes.BIDIRECTIONAL_DIJKSTRA]: 'Bidirectional Dijkstra',
  [PathfindingAlgorithmTypes.DFS]: 'DFS',
  [PathfindingAlgorithmTypes.GREEDY_BEST_FIRST_SEARCH]:
    'Greedy Best First Search',
  [PathfindingAlgorithmTypes.BIDIRECTIONAL_GREEDY_BEST_FIRST_SEARCH]:
    'Bidirectional Greedy Best First Search',
};
